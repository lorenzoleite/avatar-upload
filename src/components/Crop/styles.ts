import styled, { css } from 'styled-components';

export const StyledCrop = styled.div`
  ${({ theme }) => css`
    display: flex;
    justify-content: space-between;
    width: 55.3rem;
    height: 17.7rem;
    padding: 3.2rem;
    border-radius: 0.8rem;
    background-color: ${theme.color['gray.100']};
  `}
`;

export const ImagePreview = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    justify-content: center;
    width: 11.3rem;
    height: 11.3rem;
    border-radius: 7.25rem;
    clip-path: circle();
    margin-right: 3.2rem;
    background-color: ${theme.color['gray.200']};
  `}
`;

export const CropTitle = styled.p`
  ${({ theme }) => css`
    font-family: ${theme.font.family.inter};
    font-size: ${theme.font.size.medium};
    font-weight: ${theme.font.weight.normal};
    line-height: ${theme.font.lineHeight.medium};

    color: ${theme.color['gray.400']};

    margin-bottom: 1.3rem;
  `}
`;
