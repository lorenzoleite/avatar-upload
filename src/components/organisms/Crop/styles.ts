import styled, { css } from 'styled-components';

export const StyledCrop = styled.div`
  ${({ theme }) => css`
    display: flex;
    justify-content: space-between;
    width: 55.3rem;
    height: 17.7rem;
    padding: 3.2rem;
    border-radius: 0.8rem;
    background-color: ${theme.color['gray.100']};
  `}
`;

export const CropTitle = styled.p`
  ${({ theme }) => css`
    font-family: ${theme.font.family.inter};
    font-size: ${theme.font.size.medium};
    font-weight: ${theme.font.weight.normal};
    line-height: ${theme.font.lineHeight.medium};

    color: ${theme.color['gray.400']};

    margin-bottom: 1.3rem;
  `}
`;
