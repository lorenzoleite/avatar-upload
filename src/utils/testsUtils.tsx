import React from 'react';
import fetch from 'node-fetch';

import { ThemeProvider } from 'styled-components';
import { render, RenderResult } from '@testing-library/react';

import { theme } from '../themes/theme';

export const renderWithTheme = (children: React.ReactNode): RenderResult =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);

export const imageFile = await fetch(
  'https://images.unsplash.com/photo-1661956602868-6ae368943878?ixlib=rb-4.0.3&ixid=MnwxMjA3fDF8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=234&q=16',
)
  .then(response => response.blob())
  .then(blob => new File([blob], 'image.jpg', { type: 'image/jpeg' }));
