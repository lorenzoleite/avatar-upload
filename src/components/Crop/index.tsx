import { useState } from 'react';
import AvatarEditor from 'react-avatar-editor';

import { StyledCrop, ImagePreview, CropTitle } from './styles';

import { CloseIcon } from '../../assets/CloseIcon';
import { Flex } from '../atoms/Flex';
import { Button } from '../Button';
import { Slider } from '../Slider';

type CropProps = {
  imageFile: File;
  onSave: (imageUrl: string) => void;
  onClose: () => void;
};

export function Crop({ imageFile, onSave, onClose }: CropProps) {
  const [scale, setScale] = useState(1);
  const [editor, setEditor] = useState<AvatarEditor | null>(null);

  const handleCrop = (event: React.ChangeEvent<HTMLInputElement>) => {
    setScale(Number(event.target.value));
  };

  const handleSave = () => {
    if (editor) {
      const canvas = editor.getImageScaledToCanvas();
      canvas.toBlob(blob => {
        if (blob) {
          const newImageUrl = URL.createObjectURL(blob);
          onSave(newImageUrl);
        }
      });
    }
  };

  return (
    <StyledCrop>
      <Flex align="center">
        <ImagePreview>
          <AvatarEditor
            ref={ref => setEditor(ref)}
            image={imageFile}
            scale={scale}
            width={114}
            height={114}
            borderRadius={57}
            border={0}
          />
        </ImagePreview>
        <Flex direction="column">
          <CropTitle>Crop</CropTitle>
          <Slider type="range" id="crop-slider" min="1" max="2" step="0.1" value={scale} onChange={handleCrop} />
          <Flex direction="row" width="100%" justify="flex-end">
            <Button variant="solid" onClick={handleSave}>
              Save
            </Button>
          </Flex>
        </Flex>
      </Flex>
      <Flex align="flex-start">
        <Button variant="icon" onClick={() => onClose()}>
          <CloseIcon />
        </Button>
      </Flex>
    </StyledCrop>
  );
}
