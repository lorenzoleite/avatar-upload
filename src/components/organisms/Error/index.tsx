import { StyledError, ErrorTitle } from './styles';

import { AttentionIcon } from '../../atoms/AttentionIcon';
import { CloseIcon } from '../../atoms/CloseIcon';
import { ImagePreview } from '../../molecules/ImagePreview';
import { Button } from '../../molecules/Button';
import { Flex } from '../../molecules/Flex';

type ErrorProps = {
  messageError: string;
  onClose: () => void;
};

export function Error({ messageError, onClose }: ErrorProps) {
  return (
    <StyledError>
      <Flex align="center">
        <ImagePreview>
          <AttentionIcon />
        </ImagePreview>
        <Flex direction="column" gap="0.2rem">
          <ErrorTitle>Sorry, {messageError.toLowerCase()}.</ErrorTitle>
          <Button variant="underline" onClick={() => onClose()}>
            Try again
          </Button>
        </Flex>
      </Flex>
      <Flex align="flex-start">
        <Button variant="icon" onClick={() => onClose()}>
          <CloseIcon />
        </Button>
      </Flex>
    </StyledError>
  );
}
