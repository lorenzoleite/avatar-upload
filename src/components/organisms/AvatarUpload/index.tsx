import { useMemo, useState } from 'react';
import { useDropzone } from 'react-dropzone';

import { Dropzone, DropzoneTitle, DropzoneSubtitle, Image } from './styles';

import { theme } from '../../../themes/theme';

import { MediaIcon } from '../../atoms/MediaIcon';
import { ImagePreview } from '../../molecules/ImagePreview';
import { Flex } from '../../molecules/Flex';
import { Error } from '../Error';
import { Crop } from '../Crop';

type AvatarUploadProps = {};

const acceptStyle = {
  borderColor: theme.color['blue.200'],
};

const rejectStyle = {
  borderColor: theme.color['orange.100'],
};

export function AvatarUpload({}: AvatarUploadProps) {
  const [imageFile, setImageFile] = useState<File | null>(null);
  const [croppedImage, setCroppedImage] = useState<string | null>(null);
  const [hasError, setHasError] = useState<boolean>(false);
  const [messageError, setMessageError] = useState<string | null>(null);

  const { getRootProps, getInputProps, isDragAccept, isDragReject } = useDropzone({
    accept: { 'image/png': ['.png'], 'image/jpeg': ['.jpeg', '.jpg'] },
    maxFiles: 1,
    onDropAccepted: file => {
      setImageFile(file[0]);
    },
    onDropRejected: file => {
      setHasError(true);
      setMessageError(file[0].errors[0].message);
    },
  });

  const handleClose = () => {
    setImageFile(null);
    setCroppedImage(null);
    setHasError(false);
    setMessageError(null);
  };

  const handleSave = (imageUrl: string) => {
    setImageFile(null);
    setCroppedImage(imageUrl);
  };

  const style = useMemo(
    () => ({
      ...(isDragAccept && acceptStyle),
      ...(isDragReject && rejectStyle),
    }),
    [isDragAccept, isDragReject],
  );

  return (
    <Flex>
      {!imageFile && !hasError && (
        <Dropzone {...getRootProps({ style })} justify={croppedImage ? 'flex-start' : 'center'} role="dropzone">
          <input {...getInputProps()} data-testid="inputDropzone" />
          <Flex align="center">
            {croppedImage && (
              <ImagePreview marginRight="5rem">
                <Image src={croppedImage} alt="Image Preview" />
              </ImagePreview>
            )}
            <Flex direction="column" align="center">
              <Flex align="center" gap="1.2rem">
                <MediaIcon />
                <DropzoneTitle>Organization Logo</DropzoneTitle>
              </Flex>
              <DropzoneSubtitle>Drop the image here or click to browse.</DropzoneSubtitle>
            </Flex>
          </Flex>
        </Dropzone>
      )}
      {hasError && messageError && <Error messageError={messageError} onClose={handleClose} />}
      {imageFile && <Crop imageFile={imageFile} onSave={handleSave} onClose={handleClose} />}
    </Flex>
  );
}
