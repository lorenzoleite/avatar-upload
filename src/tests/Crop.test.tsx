import { screen, fireEvent } from '@testing-library/react';
import { describe, it, vi } from 'vitest';

import { Crop } from '../components/organisms/Crop';
import { renderWithTheme, imageFile } from '../utils/testsUtils';

describe('Crop', () => {
  it('should render the Crop', async () => {
    const onSave = vi.fn();
    const onClose = vi.fn();

    const { container } = renderWithTheme(<Crop imageFile={imageFile} onSave={onSave} onClose={onClose} />);

    expect(screen.getByText(/Crop/i)).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Save/i })).toBeInTheDocument();
    expect(screen.getByTestId(/slider/i))
      .toHaveProperty('value', '1')
      .toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
  it('should be able to change the slider value', async () => {
    const onSave = vi.fn();
    const onClose = vi.fn();

    const { container } = renderWithTheme(<Crop imageFile={imageFile} onSave={onSave} onClose={onClose} />);

    const slider = screen.getByTestId(/slider/i);

    fireEvent.change(slider, {
      target: {
        value: 1.2,
      },
    });

    expect(slider).toHaveProperty('value', '1.2').toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
