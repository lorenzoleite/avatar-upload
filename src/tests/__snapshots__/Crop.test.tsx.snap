// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Crop > should be able to change the slider value 1`] = `
<div
  class="sc-irTswW egOFGJ"
>
  <div
    class="sc-hAtEyd ebydGn"
  >
    <div
      class="sc-fbJfA fLXmBR"
    >
      <canvas
        height="114"
        style="width: 114px; height: 114px; cursor: grab;"
        width="114"
      />
    </div>
    <div
      class="sc-hAtEyd lneCAO"
      direction="column"
    >
      <p
        class="sc-csCMJt ecGFru"
      >
        Crop
      </p>
      <input
        class="sc-ktEKTO brvnIf"
        data-testid="slider"
        id="crop-slider"
        max="2"
        min="1"
        step="0.1"
        type="range"
        value="1.2"
      />
      <div
        class="sc-hAtEyd dAyymh"
        direction="row"
        width="100%"
      >
        <button
          class="sc-brKeYL hvEvOw"
        >
          Save
        </button>
      </div>
    </div>
  </div>
  <div
    class="sc-hAtEyd eeUmeC"
  >
    <button
      class="sc-brKeYL gqAyHA"
    >
      <svg
        data-testid="closeIcon"
        fill="none"
        height="20"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M11.762 10l4.113-4.112c.238-.226.375-.538.375-.888a1.254 1.254 0 00-2.137-.888L10 8.239 5.888 4.113a1.254 1.254 0 00-1.775 1.775L8.238 10l-4.113 4.113A1.207 1.207 0 003.75 15a1.254 1.254 0 002.138.887L10 11.762l4.113 4.113c.225.238.537.375.887.375a1.254 1.254 0 00.887-2.137L11.762 10z"
          fill="#677489"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`Crop > should render the Crop 1`] = `
<div
  class="sc-irTswW egOFGJ"
>
  <div
    class="sc-hAtEyd ebydGn"
  >
    <div
      class="sc-fbJfA fLXmBR"
    >
      <canvas
        height="114"
        style="width: 114px; height: 114px; cursor: grab;"
        width="114"
      />
    </div>
    <div
      class="sc-hAtEyd lneCAO"
      direction="column"
    >
      <p
        class="sc-csCMJt ecGFru"
      >
        Crop
      </p>
      <input
        class="sc-ktEKTO cSNWnz"
        data-testid="slider"
        id="crop-slider"
        max="2"
        min="1"
        step="0.1"
        type="range"
        value="1"
      />
      <div
        class="sc-hAtEyd dAyymh"
        direction="row"
        width="100%"
      >
        <button
          class="sc-brKeYL hvEvOw"
        >
          Save
        </button>
      </div>
    </div>
  </div>
  <div
    class="sc-hAtEyd eeUmeC"
  >
    <button
      class="sc-brKeYL gqAyHA"
    >
      <svg
        data-testid="closeIcon"
        fill="none"
        height="20"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M11.762 10l4.113-4.112c.238-.226.375-.538.375-.888a1.254 1.254 0 00-2.137-.888L10 8.239 5.888 4.113a1.254 1.254 0 00-1.775 1.775L8.238 10l-4.113 4.113A1.207 1.207 0 003.75 15a1.254 1.254 0 002.138.887L10 11.762l4.113 4.113c.225.238.537.375.887.375a1.254 1.254 0 00.887-2.137L11.762 10z"
          fill="#677489"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;
